README.md
![](/logo.png)

***
## What is MergeCrunch?

An application for easily download video content from Crunchyroll combining youtube-dl and mkvmerge. Build a fancy-mkv file with subtitles and fonts attached.

## Docker image

==This is the recommended way to use this application==

Find this project on DockerHub. Its latest release is working fully at March 20th, 2019

**Zero-configuration, for any operating system (Windows, MacOS and Linux). Just pull and run**

More information at https://hub.docker.com/r/beardoverflow/mergecrunch

## Features

 - Choose between softsub or hardsub (no transcoding!)
 - Choose as you want to download: individual episode, several episodies or full series
 - Choose which quality you want to get (from 240p to 1080p)
 - On softsubbing, attach only one subtitle track or all subtitles track
 - On softsubbing, attach your own fonts to the mkv file
 - On softsubbing, warn if there are fonts missing
 - Spoof your user-agent and cookies file for logging with your account

## Usage

In this section, I will illustrate how to use the application with examples:

**Basic example (argument -i):**

Input URL using argument -i. Episode will be downloaded with max resolution in the current directory.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE
```

**Playlist selection (append # character in input URL):**

For playlist selection, you must append the # character in input URL. After, use selection syntax:

- "N-M" for selecting inclusive range from N to M.
- "N" for simple selection.
- "," as separator for multiple selections.

Example #1. Select IDs items from 12 to 20.
```sh
./mergecrunch.sh -i URL_PLAYLIST_CRUNCH_HERE#12-20
```

Example #2. Select ID item 5.
```sh
./mergecrunch.sh -i URL_PLAYLIST_CRUNCH_HERE#5
```

Example #3. Select IDs items from 12 to 20 and also 2, 5, 23 to 30.
```sh
./mergecrunch.sh -i URL_PLAYLIST_CRUNCH_HERE#12-20,2,5,23-30
```

**Output file name (argument -o):**

In this example, output file will renamed as "Sket Dance 01 [CRC32_HERE].mkv"

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -o 'Sket Dance 01.mkv'
```

**CRC32 example (argument -x):**

In this example, CRC32 will be calculated and stored in the filename.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x
```

**Format example (argument -f):**

In this example, resolution will be 1280x720. Be careful with this argument, some resolution are availabled for premium users only.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p
```

Format | Description
------ | -----------
worst  | The worst resolution available (generally 240p, 360p or 480p)
240p   | 320x240 or 420x240
360p   | 480x360 or 640x360
480p   | 640x360 or 848x480
720p   | 1280x720
1080p  | 1920x1080
best   | The best resolution available (generally 480p, 720p or 1080p)

**Preferred language (argument -s) + Only one language (argument --one):**

Using a preferred language, you set a default subtitle track in your mkv. In this example, set spanish subtitle track as preferred. If you append the --one argument, then esES subtitle track will merged exclusively.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p --one -s esES
```

Language | Description
-------- | -----------
enUS     | Forces American English
esES     | Forces European Spanish
esLA     | Forces American Spanish
frFR     | Forces Français
itIT     | Forces Italiano
ptBR     | Forces Brazilian Português
ptPT     | Forces European Português
deDE     | Forces Deutsch
arME     | Forces العربية
ruRU     | Forces Русский
jaJP     | Forces 日本語

**Hardsubbing switch (argument --hard):**

If you wish download a hardsub video instead of merging a soft subtitle track, you can append the --hard argument. Require -s argument and implies --one argument.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p --hard -s esES
```

**Account access (argument --cookies and --ua):**

You must get your cookies file (to export your Crunchyroll authentication) and know which is your user agent (bypassing Cloudflare). 

In order to get your cookie file (argument --cookies or -c), I would recommend to use an extension navigator such as [cookies.txt from Chrome Store](https://chrome.google.com/webstore/detail/cookiestxt/njabckikapfpffapmjgojcnbfjonfjfg) or [cookies.txt from Firefox Add-Ons](https://addons.mozilla.org/en-US/firefox/addon/cookies-txt/).
```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p -s esES -o 'Sket Dance 01.mkv' -c cookies.txt
```

In order to get your user agent, [ask to google for getting it using the same browser where you downloaded the cookie file](https://www.google.com/search?q=what+is+my+user+agent)
```
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p -s esES -o 'Sket Dance 01.mkv' -c cookies.txt --ua 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0'
```

## For manual install without docker: Dependencies

==Manual install is not recommended way. Only for Debian an derivates. Just for testing purposes==

Install youtube-dl, python3, fontconfig, mkvmerge and rhash

For getting this dependencies, execute the classic sudo apt-get install.

```sh
sudo apt-get install youtube-dl python3 fontconfig mkvtoolnix rhash
```

Note 1. Latest version of youtube-dl on [all-in-on binary](https://ytdl-org.github.io/youtube-dl/download.html)

Note 2. Latest version of mkvmerge on [custom Bunkus' repository](https://mkvtoolnix.download/downloads.html#debian)

Fontconfig is the engine to search for fonts in your system. If the applicationg warns you about missing fonts, create a folder in your home path called ~/.fonts and put in here the missing fonts

## DEPRECATED OPTIONS

**Premium account (argument -u and -p):**

In this example, I am logging in my premium account. **Deprecated by Crunchyroll’s new verifications.** The console will prompt for username's password.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p -s esES -o 'Sket Dance 01.mkv' -u BeardOverflow
```

However, you may specific your password by command line.

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p -s esES -o 'Sket Dance 01.mkv' -u BeardOverflow -p mysecretpassword
```

**Spoof location (argument -g):**

Similar to choose your preferred language, you can spoof your location in order to download videos from foreign locations. The following example shows a spoof location to Russia and preferred language to American Spanish. **Deprecated by Crunchyroll's new verifications.**

```sh
./mergecrunch.sh -i URL_CRUNCH_HERE -x -f 720p -s esLA -g ruRU
```

## FEEDBACK, BUGS OR CONTRIBUTION

Open an issue in this repository or fork this

## LICENSE
GNU General Public License v2.0


CHANGELOG.txt
2018-03-20
	-- Rev. 24 --
		[Refactor] Show a warning about using the arguments --ua and --cookies instead of force its use
		[Docker] First dockerized version at https://hub.docker.com/r/beardoverflow/mergecrunch

2018-03-19
	-- Rev. 23 --
		[New] Argument --hard for downloading hardsub video instead of softsub tracks.
	-- Rev. 22 --
		[New] Argument --one for downloading exclusively the subtitle track specified by argument -s.

2018-03-02
	-- Rev. 21 --
		[New] Support for setting your user-agent.

2018-10-06
	-- Rev. 20 --
		[Improvement] Add pattern for trTR subtitle language.
		[Bug] Not registered subtitle patterns breaks the whole merge process. Just a warning now.
		[Bug] Bash's substitution history feature breaks any name containing the !. string.
		[Improvement] Any warning in console are yellow colored now (before it was mixing between red and yellow).

2018-09-27
	-- Rev. 19 --
		[Improvement] Spoofing your location and setting the preferred language works separately.
		[Bug] Some title with two or more consecutive spaces breaks mkvmerge.

2018-09-25
	-- Rev. 18 --
		[New] Support to spoof your location using a fake ip.
		[Bug] Some titles with rare characters as ` breaks mkvmerge.

2018-09-17
	-- Rev. 17 --
		[Bug] Streaming without subtitles cannot build to mkv.
		[Improvement] Add pattern for jaJP track language.

2018-01-02
        -- Rev. 16 --
                [Improvement] Smart selection for resolution with simple tags (worst, 360p, ..., 1080p, best).
                [Improvement] Default subtitle is forced subtitle too (because some players ignore default mark).
        -- Rev. 15 --
                [New] Support for cookies file.
        -- Rev. 14 --
                [Bug] Web scraping was failing for playlist url.
                [Improvement] Add pattern for ruRU subtitle language.

2017-01-23
	-- Rev. 13 --
		[Improvement] Supress dummy message using ls command.

2015-09-07
	-- Rev. 12 --
		[Improvement] Detecting new streaming container (mp4).

2015-08-03
	-- Rev. 11 --
		[Improvement] Warnings in mkvmerge are accepted as valid file and console output shows this situation.
	-- Rev. 10 --
		[Bug] Playlist selection gets more IDs than you request. (For example, you request 5 and it selects 5 and 50).

2015-08-02
	-- Rev. 9 --
		[Improvement] Considering style for font matching (Arial Narrow, I see you).
		[Bug] Instability for invalid links with errors due to restricted region or other reason.

2015-07-31
	-- Rev. 8 --
		[Improvement] Playlist selection.
		[Bug] Some titles with rare characters as ' breaks mkvmerge.

2015-07-30
	-- Rev. 7 --
		[New] Support for playlist URL.
		[Improvement] Add pattern for trTR subtitle language.

2015-07-28
	-- Rev. 6 --
		[Bug] Output argument (-o) accepts relative and absolute path correctly.
	-- Rev. 5 --
		[New] Support to premium users.

2015-07-27
	-- Rev. 4 --
		[Bug] Console shows an incorrect output filename if CRC32 sum had actived.
		[Bug] Breadcrumb with blackspaces in fonts text.
	-- Rev. 3 --
		[Improvement] Console output is cleaner.
	-- Rev. 2 --
		[New] Title recognition when you specific a language with argument -s
		[Improvement] Add patterns for deDE and arME subtitles language.

2015-07-26
	-- Rev. 1 --
		[New] First release. Crunchyroll streaming downloads with soft-subs and fonts attachment to mkv file.


mergecrunch.sh
#!/bin/bash

# Title:       MergeCrunch
# Description: Download from Crunchyroll and generate a pretty mkv file with all video, subtitles and fonts merged.
# Author:      José Ángel Pastrana Padilla
# Email:       japp0005@red.ujaen.es
# Last update: 2019-03-20
# Revision:    24

# DEPENDENCIES:

# - fontconfig (and enough fonts installed in your system. If any font is missing, put it inside your fonts folder ~/.fonts).
# - youtube-dl
# - mkvmerge
# - rhash (recommend, but it is not required).

# DEFAULT PARAMETERS.

# This argument sets a premium username Crunchyroll account.
# By command line, it is "-u username" or "--username username"
USERNAME="" # Default: "" (Means guest session).

# This argument sets a password for your username Crunchyroll account.
# By command line, it is "-p password" or "--password password"
PASSWORD="" # Default: "" (Means guest session if username is not specific).

# This argument sets where cookies file is stored.
# By command line, it is "-c cookies.txt" or "--cookies cookies.txt"
COOKIES="" # Default: "" (Means not using previous cookies).

# This argument computes a CRC32 sum to downloaded file and renames it in order to append the hash.
# By command line, it can be actived by "-x" or "--crc32"
CRC32="" # Default: "" (Means disabled). Change to "YES" for always active

# This argument sets video resolution.
# By command line, it is "-f value" or "--format value".
# Default: "" (Means best resolution).
# Change to other value as "worst", "240p", "360p", "480p", "720p", "1080p", "best".
FORMAT=""

# This argument sets user agent.
# By command line, it is, for example, "--ua 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0'"
USER_AGENT=""

# Only the language specified on SUB_DEFAULT (-s) will be merged
# By command line, it is, for example, "--one -s esES"
ONE_SUBTITLE=""  # Default: "" (Means disabled). Change to "YES" for always active

# Ship hard subtitle specified by SUB_DEFAULT (-s). Implies ONE_SUBTITLE
# By command line, it is, for example, "--hard -s esES"
HARD_SUBTITLE="" # Default: "" (Means disabled). Change to "YES" for always active

# For further references: ISO 639 and mkvmerge --list-languages
# This argument sets default track subtitle if your player doesn't set this field.
# By command line, use "-s value" or "--sub_default value"
# Default: "" (Means english).
# Value to "enUS" forces American English.
# Value to "esES" forces European Spanish.
# Value to "esLA" forces American Spanish.
# Value to "frFR" forces Français.
# Value to "itIT" forces Italiano.
# Value to "ptBR" forces Brazilian Português.
# Value to "ptPT" forces European Português.
# Value to "deDE" forces Deutsch.
# Value to "arME" forces العربية.
# Value to "trTR" forces Türkçe
# Value to "ruRU" forces Русский.
# Value to "jaJP" forces 日本語.
SUB_DEFAULT=""

# This argument spoofs X-Forwarded-For using a fake IP
# By command line, it is "-g US" or "--geo enUS"
# Default: "" (Means your actual IP country)
# Value to "enUS" forces USA.
# Value to "esES" forces España.
# Value to "esLA" forces México.
# Value to "frFR" forces France.
# Value to "itIT" forces Italia.
# Value to "ptBR" forces Brasil.
# Value to "ptPT" forces Brasil.
# Value to "deDE" forces Deutschland.
# Value to "arME" forces العربية.
# Value to "trTR" forces Türkiye
# Value to "ruRU" forces Россия.
# Value to "jaJP" forces 日本.
GEO_COUNTRY="" # Default: "" (Means your actual IP country)

# NEED THIS VARIABLES...
TMP_DIR="/tmp/${$}"
DEST_DIR="$(pwd)"
declare -A FORMAT_SUP
FORMAT_SUP["worst"]="worst[format_id !*= hardsub]"
FORMAT_SUP["240p"]="best[format_id !*= hardsub][height=240]"
FORMAT_SUP["360p"]="best[format_id !*= hardsub][height=360]"
FORMAT_SUP["480p"]="best[format_id !*= hardsub][height=480]"
FORMAT_SUP["720p"]="best[format_id !*= hardsub][height=720]"
FORMAT_SUP["1080p"]="best[format_id !*= hardsub][height=1080]"
FORMAT_SUP["240"]="best[format_id !*= hardsub][height=240]"
FORMAT_SUP["360"]="best[format_id !*= hardsub][height=360]"
FORMAT_SUP["480"]="best[format_id !*= hardsub][height=480]"
FORMAT_SUP["720"]="best[format_id !*= hardsub][height=720]"
FORMAT_SUP["1080"]="best[format_id !*= hardsub][height=1080]"
FORMAT_SUP["best"]="best[format_id !*= hardsub]"
FORMAT_SUP["hard-worst"]="worst[format_id *= hardsub]"
FORMAT_SUP["hard-240p"]="best[format_id *= hardsub][height=240]"
FORMAT_SUP["hard-360p"]="best[format_id *= hardsub][height=360]"
FORMAT_SUP["hard-480p"]="best[format_id *= hardsub][height=480]"
FORMAT_SUP["hard-720p"]="best[format_id *= hardsub][height=720]"
FORMAT_SUP["hard-1080p"]="best[format_id *= hardsub][height=1080]"
FORMAT_SUP["hard-240"]="best[format_id *= hardsub][height=240]"
FORMAT_SUP["hard-360"]="best[format_id *= hardsub][height=360]"
FORMAT_SUP["hard-480"]="best[format_id *= hardsub][height=480]"
FORMAT_SUP["hard-720"]="best[format_id *= hardsub][height=720]"
FORMAT_SUP["hard-1080"]="best[format_id *= hardsub][height=1080]"
FORMAT_SUP["hard-best"]="best[format_id *= hardsub]"
declare -A SUB_LANG
SUB_LANG["enUS","tag"]="eng"
SUB_LANG["enUS","cty"]="English (USA)"
SUB_LANG["enUS","geo"]="US"
SUB_LANG["esES","tag"]="spa"
SUB_LANG["esES","cty"]="Español (España)"
SUB_LANG["esES","geo"]="ES"
SUB_LANG["esLA","tag"]="spa"
SUB_LANG["esLA","cty"]="Español (México)"
SUB_LANG["esLA","geo"]="LA"
SUB_LANG["frFR","tag"]="fre"
SUB_LANG["frFR","cty"]="Français (France)"
SUB_LANG["frFR","geo"]="FR"
SUB_LANG["itIT","tag"]="ita"
SUB_LANG["itIT","cty"]="Italiano (Italia)"
SUB_LANG["itIT","geo"]="IT"
SUB_LANG["ptBR","tag"]="por" # Should be bzs according to ISO-639-3, but mkvmerge only supports ISO-639-1/2
SUB_LANG["ptBR","cty"]="Português (Brasil)"
SUB_LANG["ptBR","geo"]="BR"
SUB_LANG["ptPT","tag"]="por"
SUB_LANG["ptPT","cty"]="Português (Portugal)"
SUB_LANG["ptPT","geo"]="PT"
SUB_LANG["deDE","tag"]="ger"
SUB_LANG["deDE","cty"]="Deutsch (Deutschland)"
SUB_LANG["deDE","geo"]="DE"
SUB_LANG["arME","tag"]="ara"
SUB_LANG["arME","cty"]="العربية"
SUB_LANG["arME","geo"]="ME"
SUB_LANG["trTR","tag"]="tur"
SUB_LANG["trTR","cty"]="Türkçe (Türkiye)"
SUB_LANG["trTR","geo"]="TR"
SUB_LANG["ruRU","tag"]="rus"
SUB_LANG["ruRU","cty"]="Русский (Россия)"
SUB_LANG["ruRU","geo"]="RU"
SUB_LANG["jaJP","tag"]="jap"
SUB_LANG["jaJP","cty"]="日本語 (日本)"
SUB_LANG["jaJP","geo"]="JP"

# Turn off substitution history
# Sad example: echo "blabla!.mkv"
set +H

# A SIMPLES FUNCTIONS FOR COLOURED OUTPUT TEXT
function greencon {
	echo -e "\e[1m\e[92m${*}\e[0m"
}
function redcon {
	echo -e "\e[1m\e[91m${*}\e[0m"
}
function yellowcon {
	echo -e "\e[1m\e[93m${*}\e[0m"
}


# HANDLE SIGNAL
function handlesignal {
	if [ -d "${TMP_DIR}" ]
	then
		rm -r "${TMP_DIR}"
	fi
	if [ -f "${TMP_FILE}" ]
	then
		rm "${TMP_FILE}"
	fi
	if [ ${#} -eq 0 ]
	then
		redcon "Signal received! Abort all and exit!"
	else
		redcon "${*}"
	fi
	exit -1
}
trap handlesignal SIGHUP SIGINT SIGTERM


# CHECKING...
if [ ${#} -eq 0 ]
then
	echo "Error. MergeCrunch needs a input URL as parameter."
	echo "Syntaxis:"
	echo "	${0} -i URL [-f format] [-s force] [-c] [-o output]"
	exit -1
fi

if [ -z "$(which fc-list)" ]
then
	echo "Please, first install fontconfig package."
	echo "Run 'sudo apt-get install fontconfig' and try again."
	exit -1
fi

if [ -z "$(which youtube-dl)" ]
then
	echo "Please, first install youtube-dl package."
	echo "Run 'sudo apt-get install youtube-dl' and try again."
	exit -1
fi

if [ -z "$(which mkvmerge)" ]
then
	echo "Please, first install mkvtoolnix package."
	echo "Run 'sudo apt-get install mkvtoolnix' and try again."
	exit -1
fi


# PARSING...
while [ ${#} -gt 0 ]
do
	case "${1}" in
		-i|--input)
			INPUT="${2}"
			shift
		;;
		-o|--output)
			if [ "${2:0:1}" = "/" ]
			then
				OUTPUT="${2}"
			else
				OUTPUT="${DEST_DIR}/${2}"
			fi
			if [ -z "$(echo ${OUTPUT##*/} | grep "\.")" ] # Output given is a directory.
			then
				DEST_DIR="$(readlink -f "${OUTPUT}")"
				unset OUTPUT
			fi
			shift
		;;
		-u|--username)
			USERNAME="${2}"
			shift
		;;
		-p|--password)
			PASSWORD="${2}"
			shift
		;;
		-c|--cookies)
			COOKIES="$(readlink -f "${2}")"
			sed -i 's/^#HttpOnly_//' "${COOKIES}"
			shift
		;;
		-x|--crc32)
			if [ -z "$(which rhash)" ]
			then
				echo "Using ${1} argument, it needs rhash installed."
				echo "Run 'sudo apt-get install rhash' and try again."
				exit -1
			fi
			CRC32="YES"
		;;
		-f|--format)
			if [ -z "${FORMAT_SUP["${2,,}"]}" ]
			then
				echo "Error. Format specification for ${2} not found."
			        echo "Edit this script and add it in FORMAT_SUP array."
			        exit -1
			fi
			FORMAT="${2,,}"
			shift
		;;
		-s|--sub_default)
			if [ -z "${SUB_LANG["${2}","tag"]}" ]  || [ -z "${SUB_LANG["${2}","cty"]}" ]
			then
				echo "Error. Subtitle specification for ${2} not found."
				echo "Edit this script and add it in SUB_LANG array."
				exit -1
			fi
			SUB_DEFAULT="${2}"
			shift
		;;
		-g|--geo)
			if [ -z "${SUB_LANG["${2}","geo"]}" ]
			then
				echo "Error. Geolocation specification for ${2} not found."
				echo "Edit this script and add it in SUB_LANG array."
				exit -1
			fi
			GEO_COUNTRY="${2}"
			shift
		;;
		--ua|--user-agent)
			USER_AGENT="${2}"
			shift
		;;
		--one)
			ONE_SUBTITLE="YES"
		;;
		--hard)
			HARD_SUBTITLE="YES"
		;;
		*)
			echo "Error. Unexpected argument: ${1}"
			exit -1
		;;
	esac
	shift
done
if [ -n "${USERNAME}" ]
then
	USERNAME="-u ${USERNAME}"
fi
if [ -n "${PASSWORD}" ]
then
	PASSWORD="-p ${PASSWORD}"
fi
if [ -n "${COOKIES}" ]
then
	WGET_COOKIES="--load-cookies \"${COOKIES}\""
	COOKIES="--cookies '${COOKIES}'"
else
	yellowcon "Warning. Crunchyroll will enable future restrictions soon. If there is any problem while downloading, try to specify a cookies file (for example, -c cookies.txt). More info at README.md"
fi
if [ -n "${HARD_SUBTITLE}" ] && [ -z "${SUB_DEFAULT}" ]
then
	echo "Error. If you use --hard option, then you must specify the default subtitle language using -s option"
	exit -1
fi
if [ -n "${FORMAT}" ]
then
	if [ -n "${HARD_SUBTITLE}" ]
	then
		FORMAT="-f '${FORMAT_SUP["hard-${FORMAT}"]}[format_id *= -${SUB_DEFAULT}-]'"
	else
		FORMAT="-f '${FORMAT_SUP["${FORMAT}"]}'"
	fi
else
	if [ -n "${HARD_SUBTITLE}" ]
	then
		FORMAT="-f '[format_id *= -${SUB_DEFAULT}-]'"
	fi
fi
if [ -n "${SUB_DEFAULT}" ]
then
	HEADER="--add-header Accept-Language:${SUB_DEFAULT}"
fi
if [ -n "${GEO_COUNTRY}" ]
then
	GEO_COUNTRY="--geo-bypass-country ${SUB_LANG["${GEO_COUNTRY}","geo"]}"
	FAKE_IP="$(youtube-dl -s -v ${GEO_COUNTRY} "${INPUT}" 2>&1 | grep -m1 'fake IP' | xargs | cut -d' ' -f5)"
	WGET_GEO_COUNTRY="--header \"X-Forwarded-For:${FAKE_IP}\""
	yellowcon "Using fake ip: $FAKE_IP"
fi
if [ -n "${USER_AGENT}" ]
then
	WGET_USER_AGENT="--user-agent \"${USER_AGENT}\""
	USER_AGENT="--user-agent '${USER_AGENT}'"
else
	yellowcon "Warning. Crunchyroll will enable future restrictions soon. If there is any problem while downloading, try to specify an user-agent (for example, --ua 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0'). More info at README.md"
fi
if [ -n "${ONE_SUBTITLE}" ]
then
	if [ -z "${SUB_DEFAULT}" ]
	then
		echo "Error. If you use --one option, then you must specify the default subtitle language using -s option"
		exit -1
	fi
fi

# MAIN...
# Check playlist...
TMP_FILE="/tmp/${$}-check"
eval wget \"${INPUT}\" ${WGET_USER_AGENT} ${WGET_COOKIES} ${WGET_GEO_COUNTRY} -qO \"${TMP_FILE}\"
if [ -z "$(cat "${TMP_FILE}" | grep '"type":"error"')" ] # Check URL errors
then
if [ -z "$(cat "${TMP_FILE}" | grep 'link rel="index"')" ] # Input URL is a playlist
then
	greencon "[Analyze] INPUT URL IS A PLAYLIST. ENQUEUING..."
	if [ -n "${OUTPUT}" ]
	then
		handlesignal "Output argument must be a directory when input argument is a playlist URL."
	fi
	SEGMENT="${INPUT##*#}"
	if [ -n "${SEGMENT}" ] && [ "${INPUT}" != "${SEGMENT}" ]
	then
		while read -r line
		do
			if [ -n "$(echo "${line}" | grep "-")" ]
			then
				SELECTION+="$(seq -s " " ${line/-/ }) "
			else
				SELECTION+="${line} "
			fi
		done <<< "$(echo -e "${SEGMENT//,/\\n}")"
		greencon "---> SELECTED ID ITEMS FROM THIS PLAYLIST: ${SELECTION} <---"
	fi
	c=0
	while read -r line
	do
		c=$((c+1))
		inputs+=("http://www.crunchyroll.com${line}")
		if [ -z "${SELECTION}" ] || [ -n "$(echo -e "${SELECTION// /\\n}" | grep "^${c}$")" ]
		then
			echo -en "\e[1m[Selected] \e[42m"
		else
			echo -n "[Not Selected] "
		fi
		echo -e "ID: ${c}, URL: http://www.crunchyroll.com${line}\e[0m"
	done <<< "$(tac "${TMP_FILE}" | grep -A1 --no-group-separator 'portrait-element block-link titlefix episode' | grep -v 'portrait-element block-link titlefix episode' | cut -d'"' -f2)"
else
	greencon "[Analyze] INPUT URL IS A SIMPLE URL. ENQUEUING..."
	inputs+=("${INPUT}")
	echo "${INPUT} ready!"
fi
else
	eval youtube-dl -s ${USERNAME} ${PASSWORD} ${COOKIES} ${USER_AGENT} ${FORMAT} ${HEADER} ${GEO_COUNTRY} ${INPUT}
	handlesignal "Exitting due to an error in input URL..."
fi
rm -r "${TMP_FILE}"

# LOOP FOR PLAYLIST
c=0
for INPUT in "${inputs[@]}"
do
	c=$((c+1))
	if [ -z "${SELECTION}" ] || [ -n "$(echo -e "${SELECTION// /\\n}" | grep "^${c}$")" ]
	then
	(
	# A line extra for to format console output :P
	echo 

	# Prepare temp dir
	mkdir -p "${TMP_DIR}"
	cd "${TMP_DIR}"

	# Get Crunchyroll file
	greencon "STEP 0. QUEUE INPUT URL #${c}."
	echo "${INPUT}"
	greencon "STEP 1. GET CRUNCHYROLL STREAMING DOWNLOAD."
	echo youtube-dl --no-warnings --no-continue --no-part --all-subs ${USERNAME} ${PASSWORD} ${COOKIES} ${USER_AGENT} ${FORMAT} ${HEADER} ${GEO_COUNTRY} ${INPUT}
	eval youtube-dl --no-warnings --no-continue --no-part --all-subs ${USERNAME} ${PASSWORD} ${COOKIES} ${USER_AGENT} ${FORMAT} ${HEADER} ${GEO_COUNTRY} ${INPUT}
	if [ ${?} -ne 0 ]
	then
		handlesignal "Failed to get streaming! Exitting..."
	fi

	# Prepare output filename
	DL_NAME="$(ls *.flv *.mp4 2>/dev/null)"
	if [ -z "${OUTPUT}" ]
	then
		OUTPUT="${DEST_DIR}/${DL_NAME%-*}.mkv"
	fi

	# Prepare subtitles downloaded for will merge to output file
	greencon "STEP 2. CHECKING AVAILABLE SUBTITLES."
	if ls *.ass >/dev/null 2>/dev/null
	then
		if [ -n "${HARD_SUBTITLE}" ]
		then
			yellowcon "Hard subtitle enabled. Skip this step."
		else
		for each in *.ass
		do
			sl=${each%.ass}
			sl=${sl##*.}
			if [ -z "${SUB_LANG["${sl}","tag"]}" ] || [ -z "${SUB_LANG["${sl}","cty"]}" ]
			then
				yellowcon "WARNING: UNABLE TO UNDERSTAND <<< ${sl} >>>. EDIT THIS SCRIPT AND ADD IT IN SUB_LANG ARRAY. WHILE I WILL IGNORE IT."
				continue
			fi
			unset def
			if [ "${SUB_DEFAULT}" = "${sl}" ]
			then
				def="--default-track 0:yes --forced-track 0:yes"
			elif [ -n "${ONE_SUBTITLE}" ]
			then
				continue
			fi
			echo "Found ${SUB_LANG["${sl}","cty"]} subtitle!... Ready."
			SUB_MKV+=("--language \"0:${SUB_LANG["${sl}","tag"]}\" --track-name \"0:${SUB_LANG["${sl}","cty"]}\" ${def} '(' \"${each}\" ')'")
		done
		fi
	else
		yellowcon "Not found any subtitle track."
	fi

	# Searching for fonts attachments
	greencon "STEP 3. CHECKING AVAILABLES ATTACHMENT FONTS TEXT."
	if ls *.ass >/dev/null 2>/dev/null
	then
		if [ -n "${HARD_SUBTITLE}" ]
		then
			yellowcon "Hard subtitle enabled. Skip this step."
		else
		for each in *.ass
		do
			while read -r line
			do
			        fonts+=("${line}")
			done <<< "$(cat "${each}" | grep "^Style:" | cut -d"," -f2 | sort -u)"
		done

		while read -r line
		do
			queryfc="$(fc-match "${line}")"
			foundName="$(echo ${queryfc} | cut -d'"' -f2)"
			foundStyle="$(echo ${queryfc} | cut -d'"' -f4)"
			if [ "${foundName,,}" = "${line,,}" ] || [ "$(echo "${foundName,,} ${foundStyle,,}" | xargs)" = "$(echo "${line,,}" | xargs)" ]
			then
			        echo "Found ${line} font!... Ready."
			else
				yellowcon "WARNING: REQUEST FONT <<< ${line} >>> IS NOT INSTALLED IN YOUR SYSTEM. PLEASE, INSTALL IT AND TRY AGAIN. WHILE I WILL BE USE <<< ${foundName} >>>."
			fi
			FONT_MKV+=("--attach-file \"$(fc-match -v "${line}" | grep "file:" | cut -d'"' -f2)\"")
		done <<< "$(printf "%s\n" "${fonts[@]}" | sort -u)"
		fi
	else
		yellowcon "Not required any attachment."
	fi

	# Launch mkvmerge
	greencon "STEP 4. TIME FOR MERGING ALL TO MKV FILE."
	MKVCOMMAND="mkvmerge --output \"${OUTPUT}\" --language 0:jpn --track-name \"0:${DL_NAME%-*}\" --language 1:jpn --track-name \"1:${DL_NAME%-*}\" '(' \"${DL_NAME}\" ')' ${SUB_MKV[*]} ${FONT_MKV[*]} --title \"${DL_NAME%-*}\" -q"
	eval "${MKVCOMMAND//\`/\\\`}"
	case "${?}" in
		0) echo "Merge completed!";;
		1) yellowcon "Merge completed with WARNINGS!";;
		2) handlesignal "Merge failed! Exitting...";;
	esac

	# Create sum if crc32 is actived by parameter
	if [ -n "${CRC32}" ]
	then
		greencon "STEP 5. CALCULATING CRC32 HASH."
		echo "CRC32 sum value is: $(rhash --simple -Ce "${OUTPUT}" | cut -d" " -f1)"
	fi

	# Final message
	greencon "FINISH!\e[0m\e[1m Output file generated: $(ls -t "${OUTPUT%.mkv}"* | head -n1)"

	# Delete temp dir
	rm -r "${TMP_DIR}"
	)
	fi
done

